src/librustc_mir/transform/uniform_array_move_out.rs

```diff
@@ -151,3 +152,177 @@ impl<'a, 'tcx> UniformArrayMoveOutVisitor<'a, 'tcx> {
+                                let opt_index_and_place = try_get_item_source(local_use, mir);
+                                if local_use.use_count == 2 && opt_index_and_place.is_some() {
+                                    let (index, src_place) = opt_index_and_place.unwrap();
+                                    return Some((local_use, index, src_place));
+                                }
+                            }
+                            None
+                        }).collect();
+
+                        let opt_src_place = items.first().and_then(|x| *x).map(|x| x.2);
+                        let opt_size = opt_src_place.and_then(|src_place| {
+                            let src_ty = src_place.ty(mir, tcx).to_ty(tcx);
+                            if let ty::TyArray(_, ref size_o) = src_ty.sty {
+                                size_o.val.to_const_int().and_then(|v| v.to_u64())
+                            } else {
+                                None
+                            }
+                        });
+
+                        check_and_patch(*candidate, &items, opt_size, &mut patch, dst_place);
+                    }
+                }
+            }
+        }
+        patch.apply(mir);
+    }
+}
+
+fn try_get_item_source<'a, 'tcx>(local_use: &LocalUse,
+                                 mir: &'a Mir<'tcx>) -> Option<(u32, &'a Place<'tcx>)> {
+    if let Some(location) = local_use.first_use {
+        let block = &mir[location.block];
+        if block.statements.len() > location.statement_index {
+            let statement = &block.statements[location.statement_index];
+            if let StatementKind::Assign(
+                Place::Local(_),
+                Rvalue::Use(Operand::Move(Place::Projection(box PlaceProjection{
+                    ref base, elem: ProjectionElem::ConstantIndex{
+                        offset, min_length: _, from_end: false}})))) = statement.kind {
+                return Some((offset, base))
+            }
+        }
+    }
+    None
+}
+
+fn check_and_patch<'tcx>(candidate: Location,
+                         items: &Vec<Option<(&LocalUse, u32, &Place<'tcx>)>>,
+                         opt_size: Option<u64>,
+                         patch: &mut MirPatch<'tcx>,
+                         dst_place: &Place<'tcx>) {
+    let opt_src_place = items.first().and_then(|x| *x).map(|x| x.2);
+
+    if opt_size.is_some() && items.iter().all(
+        |l| l.is_some() && l.unwrap().2 == opt_src_place.unwrap()) {
+
+        let (min, max, sum) = items.iter().fold(
+            (u32::max_value(), 0, 0),
+            |ac, x| {
+                let x = x.unwrap();
+                (cmp::min(ac.0, x.1), cmp::max(ac.1, x.1), ac.2 + x.1)
+            });
+
+        // checks that all items from [mix..max] are exist,
+        // expects all index are unique, borrowck have to check it
+        if sum != (min + max) * (max - min + 1) / 2 {
```

Comment:
I didn't think hard about this math; I guess it's probably correct -- but wouldn't it maybe be easier to check in the "source"? i.e., when we see a aggregate like `[A, B, C]`, couldn't we check that B is from the place "after A", and C "after B", and so forth? That feels like it'd be a bit easier

Reply:
That is, rather than gathering up a `Vec<Option<...>>`, we could check that everything matches and just know that we are pasting N consecutive items starting from A, right?